torch/lib/THD/base/data_channels/Store.cpp

```diff
@@ -0,0 +1,205 @@
+#include "Store.hpp"
+#include "../ChannelUtils.hpp"
+
+#include <poll.h>
+#include <system_error>
+
+namespace thd {
+
+using size_type = uint64_t;
+using store_type = std::unordered_map<std::string, std::vector<char>>;
+
+static void oracle_daemon();
+static bool query(int socket);
+static bool check(const store_type& store,
+                  const std::vector<std::string>& keys);
+static void wakeup(int socket);
+
+
+enum query_type : uint8_t {
+  set,
+  get,
+  wait,
+  finish,
+  stop_waiting
+};
+ 
+Store::Store()
+    : rank(load_rank_env()) {
+  // Master runs the oracle
+  if (rank == 0) {
+    std::thread oracle(oracle_daemon);
+    oracle.detach();
+  }
+  // everyone introduces himself to the oracle
+  std::tie(master_addr, master_port) = load_worker_env();
```

Comment:
This does not seem to be right because master does not have this env variables. First thing you should change is add `else` (for workers). Second thing is that you can master to oracle_deamon by:
```
port_type port;
std::tie(port, std::ignore) = load_master_env();
socket = connect("localhost", port); // not sure if "localhost" or "127.0.0.1"
```